### PR TITLE
drivers: clock_control: clock_control_on is blocking

### PR DESCRIPTION
**Pull request by mbolivar**
Document this behavior. This partially addresses #20708.

### CODE REVIEW

File: include/drivers/clock_control.h [hunk 1]
```diff
@@ -103,9 +103,17 @@ struct clock_control_driver_api {
 
 /**
  * @brief Enable the clock of a sub-system controlled by the device
- * @param dev Pointer to the device structure for the clock controller driver
- * 	instance
- * @param sys A pointer to an opaque data representing the sub-system
+ *
+ * This function is not blocking: the clock may or may not be enabled
```
Comment (tbursztyka) [inline include/drivers/clock_control.h]: That's actually wrong. See my comment on your issue. 

Common pattern being: functions ending with _async are non blocking, others are.
The documentation was missing at that time, yes. But these were meant to be blocking and should be blocking or then they would be aliases to _async counterpart with a cb set as NULL... so they would be useless.

Comment (mbolivar) [reply include/drivers/clock_control.h]: Thanks for responding. I asked you on slack yesterday what you meant but got no answer :).

OK, so then the situation is that all the clock drivers which aren't blocking are buggy. Would you agree?

Comment (mbolivar) [reply include/drivers/clock_control.h]: >  functions ending with _async are non blocking, others are.

BTW, I fully agree with this, but nobody who responded so far seemed to, so I am glad we are on the same page.

Comment (mbolivar) [reply include/drivers/clock_control.h]: > Would you agree?

I see from your comment on the issue that you do agree. Thanks!

**Comment by mbolivar** [comment]
@tbursztyka I've rewritten this with the opposite sense. Please review and I guess we'll discuss at the next API meeting.

**Review by galak — APPROVED** [review]

**Review by henrikbrixandersen — APPROVED** [review]

File: include/drivers/clock_control.h (hunk 1 above)
Comment (pabigot) [reply include/drivers/clock_control.h]: > > functions ending with _async are non blocking, others are.
> 
> BTW, I fully agree with this, but nobody who responded so far seemed to, so I am glad we are on the same page.

I also think this should be true, but see https://github.com/zephyrproject-rtos/zephyr/issues/15703#issuecomment-487470785.  "Non-blocking" is subject to conditions, at least for some API.

Comment (mbolivar) [reply include/drivers/clock_control.h]: That comment seems to indicate that "non-blocking" is not the same as "queued", which makes sense.

File: include/drivers/clock_control.h [hunk 2]
```diff
@@ -103,9 +103,16 @@ struct clock_control_driver_api {
 
 /**
  * @brief Enable the clock of a sub-system controlled by the device
- * @param dev Pointer to the device structure for the clock controller driver
- * 	instance
- * @param sys A pointer to an opaque data representing the sub-system
+ *
+ * This function is blocking: on success, the clock is enabled and
+ * ready when it returns. Use clock_control_async_on() for
```
Comment (nordic-krch) [inline include/drivers/clock_control.h]: `@ref clock_control_async_on()`?

File: include/drivers/clock_control.h [hunk 3]
```diff
@@ -103,9 +103,16 @@ struct clock_control_driver_api {
...
+ * non-blocking operation.
+ *
+ * @param dev Device structure whose driver controls the clock.
+ * @param sys Opaque data representing the clock.
```
Comment (nordic-krch) [inline include/drivers/clock_control.h]: It's not clear what it is actually. It's probably something that is vendor dependent.

**Review by nordic-krch — COMMENTED** [review]
LGTM, 2 minor comments

File: include/drivers/clock_control.h (hunk 1 above)
Comment (tbursztyka) [reply include/drivers/clock_control.h]: > Thanks for responding. I asked you on slack yesterday what you meant but got no answer :).

Sorry I missed the message/notification 

> OK, so then the situation is that all the clock drivers which aren't blocking are buggy. Would you agree?

Yes, we are on same page.

**Review by tbursztyka — APPROVED** [review]

File: include/drivers/clock_control.h [hunk 4]
```diff
@@ -103,9 +103,16 @@ struct clock_control_driver_api {
 
 /**
  * @brief Enable the clock of a sub-system controlled by the device
- * @param dev Pointer to the device structure for the clock controller driver
- * 	instance
- * @param sys A pointer to an opaque data representing the sub-system
+ *
+ * This function is blocking: on success, the clock is enabled and
```
Comment (nordic-krch) [inline include/drivers/clock_control.h]: How should it behave when called from interrupt context? Do we forbid that or it should be blocking wait (with `k_cpu_idle`)?

**Comment by ioannisg** [comment]
@mbolivar could you address the review by @nordic-krch ?

File: include/drivers/clock_control.h (hunk 4 above)
Comment (mbolivar) [reply include/drivers/clock_control.h]: Yep, that is the 64 thousand dollar question.

We have a few choices which I was planning to discuss at API meeting.

Comment (henrikbrixandersen) [reply include/drivers/clock_control.h]: Why would it be any different from other blocking functions in this regard?

Comment (mbolivar) [reply include/drivers/clock_control.h]: Looks like we won't get to discuss in the API meeting. Here are some options:

1. busy-wait in ISR context, sleep in thread context
2. busy-wait no matter the context
3. fail in ISR, sleep in thread
4. ???

I'm not sure if there are other contexts we need to worry about. I don't know how (meta IRQs)[https://docs.zephyrproject.org/latest/reference/kernel/scheduling/index.html?highlight=meta%20irq#meta-irq-priorities] work, for example.

Considering that some clocks take a long time to start up and stabilize, sleeping the thread seems desirable, if a bit more complex.

I'd like to discuss in the meeting to get as much feedback as possible from people who are interested.

Comment (mbolivar) [reply include/drivers/clock_control.h]: @carlescufi points out this is a special case of https://github.com/zephyrproject-rtos/zephyr/issues/6184.

Comment (nordic-krch) [reply include/drivers/clock_control.h]: we can have `k_cpu_idle()` in ISR and semaphore in thread context.

Personally, I would prefer that function works in both context. I wouldn't like to have `_isr`, rather function should internally handle it.

File: include/drivers/clock_control.h (hunk 3 above)
Comment (mbolivar) [reply include/drivers/clock_control.h]: I don't understand this comment. That's what "opaque data" means, no?

File: include/drivers/clock_control.h (hunk 4 above)
Comment (mbolivar) [reply include/drivers/clock_control.h]: Updated following discussion in the API meeting. The intent has apparently been that unless specifically marked, all driver APIs are only safe to call from threads.

File: include/drivers/clock_control.h (hunk 3 above)
Comment (mbolivar) [reply include/drivers/clock_control.h]: I've updated the clock disable comment to use the same language

**Review by pabigot — APPROVED** [review]
I'm fine with this as a documentation fix for 2.1.  As discussed in #20806 and suggested in https://github.com/zephyrproject-rtos/zephyr/pull/20796#pullrequestreview-321967523 we should revisit this API in terms of naming, description of behavior, and functionality.

**Review by nordic-krch — APPROVED** [review]

**Pull request merged**